pkg/providers/tinkerbell/tinkerbell.go

```diff
@@ -286,3 +291,66 @@ func (p *Provider) ChangeDiff(currentSpec, newSpec *cluster.Spec) *types.Compone
 func (p *Provider) InstallCustomProviderComponents(ctx context.Context, kubeconfigFile string) error {
 	return nil
 }
+
+// HardwareSpec returns the hardware yaml containing configurations for the hardware in the cluster.
+func (p *Provider) HardwareSpec() []byte {
+	return p.hardwareSpec
+}
+
+// generateHardwareSpec reads the hardware information from the available cluster, generates a yaml that can be submitted to a kubernetes cluster
+// and caches it on the Provider.
+func (p *Provider) generateHardwareSpec(ctx context.Context, cluster *types.Cluster) error {
```

Comment:
I think we should return the spec and then cache it in the caller of this function.